### PR TITLE
Fuzzy-search emojis, insert real unicode symbols option, read config from .czrc

### PR DESCRIPTION
**Pull request by nshtg**
- I added support for fuzzy searching the emoji list. Scrolling using my arrow keys got quite exhausting.
- I added support for an option called `symbol`. If it is set to `true`, we insert the unicode emoji instead of its textual representation.
- If there is no `package.json` with options for cz-emoji, look in the `.czrc`.

My `.czrc` looks like this:

```json
{
  "config": {
    "cz-emoji": {
      "symbol": true
    }
  },
  "path": "cz-emoji"
}
```

I did not add anything to the README (yet) as I'm not really good at the whole write-a-readme-thingy. And please excuse my ugly javascript, I use coffeescript 99.9999% of the time so my javascript is basic at most.

Example:

### CODE REVIEW

**Comment by nshtg** [comment]
Oh yes, this includes my previous PR aswell, so merging this will close #13.

**Comment by nshtg** [comment]
I added the missing code to include the closed issues in the commit message.

File: index.js [hunk 1]
```diff
@@ -76,7 +110,17 @@ function format(answers) {
   // wrap body at 100
   const body = wrap(answers.body, 100)
 
-  return (head + '\n\n' + body)
+  let commitmsg = head + '\n\n' + body
+
+  const issues = answers.issues.match(/#\d+/g)
```
Comment (ngryman) [inline index.js]: Ok so I guess you want to trigger Github automatic issue closing, that's a cool addition 👍

Just a small tweak:
```js
const footer = answers.issues.match(/#\d+/g)
  .map(issue => `Closes ${issue}`)
  .join('\n')

return [head, body, footer].join('\n\n')
```

File: index.js [hunk 2]
```diff
@@ -86,7 +130,9 @@ function format(answers) {
  */
 module.exports = {
   prompter: function(cz, commit) {
+    cz.prompt.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'))
     readPkg()
+      .then(loadConfig)
```
Comment (ngryman) [inline index.js]: After the `loadConfig` change this becomes:

```js
loadConfig()
  .then(createQuestions)
  .then(cz.prompt)
  .then(format)
```

File: index.js [hunk 3]
```diff
@@ -1,46 +1,80 @@
...
+      config = czrc.config['cz-emoji']
+    } catch (e) {
+      config = {}
+    }
+
+    config.types = config.types || types
+    config.symbol = config.symbol || false
+    return config
+  }
+}
```
Comment (ngryman) [inline index.js]: Small tweak to get rid of `lodash` and refactor a little. `readPkg` should be part of the `loadConfig` functions (see the bottom of the file).

```js
const defaultConfig = {
  types,
  symbol: false
}
```

```js
function loadConfig() {
  return readPkg()
    .then(({ pkg }) => {
      const config = (pkg && pkg.config && pkg.config['cz-emoji'])
      if (config) return config

      return new Promise((resolve, reject) => {
        fs.readFile(homeDir('.czrc'), 'utf8', (err, content) => {
          if (err) reject(err)
          resolve(JSON.parse(content))
        })
      })
    })
    .then(config => (
      Object.assign({}, defaultConfig, config)
    ))
}
```

File: index.js [hunk 4]
```diff
@@ -1,46 +1,80 @@
...
- * @param {Object} res Result of the `readPkg` returned promise
+ * @param {Object} config Result of the `loadConfig` returned promise
  * @return {Array} Return an array of `inquier.js` questions
  * @private
  */
-function createQuestions(res) {
-  const config = res.pkg.config || {}
-  const emojiConfig = config['cz-emoji'] || {}
+function createQuestions(config) {
+  const choices = getEmojiChoices(config.types, config.symbol)
```
Comment (ngryman) [inline index.js]: Let's pass the config directly: `getEmojiChoices(config)`

File: index.js [hunk 5]
```diff
@@ -1,46 +1,80 @@
...
 const truncate = require('cli-truncate')
 const wrap = require('wrap-ansi')
 const pad = require('pad')
+const fuse = require('fuse.js')
+const Promise = require('bluebird')
+const homeDir = require('home-dir')
 const types = require('./lib/types')
 
-function getEmojiChoices(types) {
+function getEmojiChoices(types, symbol) {
```
Comment (ngryman) [inline index.js]: Let's get the config directly and destructure it:
`function getEmojiChoices({ types, symbol }) {`

File: package.json [hunk 6]
```diff
@@ -19,7 +19,13 @@
     "emoji"
   ],
   "dependencies": {
+    "bluebird": "^3.5.1",
     "cli-truncate": "^1.0.0",
+    "fs-extra": "^4.0.2",
+    "fuse.js": "^3.2.0",
+    "home-dir": "^1.0.0",
+    "inquirer-autocomplete-prompt": "^0.12.0",
+    "lodash": "^4.17.4",
```
Comment (ngryman) [inline package.json]: `bluebird`, `fs-extra` and `lodash` 🗑

File: index.js [hunk 7]
```diff
@@ -1,46 +1,80 @@
...
+      type: 'autocomplete',
       name: 'type',
       message: "Select the type of change you're committing:",
-      choices: getEmojiChoices(emojiConfig.types || types)
+      source: (ansersSoFar, query) => {
+        if (!query) {
+          return Promise.resolve(choices)
+        }
+        return Promise.resolve(fuzzy.search(query))
+      }
```
Comment (ngryman) [inline index.js]: I think everything is synchronous here so we can simply do:
```js
source: (answersSoFar, query) => query ? fuzzy.search(query) : choices
```

Comment (nshtg) [reply index.js]: inquirer-autocomplete-prompt expects a promise:
> Source will be called with previous answers object and the current user input each time the user types, it must return a promise.

File: index.js (hunk 1 above)
Comment (nshtg) [reply index.js]: If no matches are found match returns `null` and not an empty array. If the footer is empty we would have 2 additional newlines appended at the end, or do they get removed automatically later?

**Comment by nshtg** [comment]
Most things are implemented, except the changes in the format function.

**Comment by nshtg** [comment]
Another fix commit, your changes did not actually load the cz-emoji subsection in the czrc.

I added support to fuzzy match the unicode 'code'. I tend to type hammer or art instead of refactoring, this make it quite a bit easier - and since its fuzzy the closer you get to feature the more it will match "that" key.

**Comment by nshtg** [comment]
Also when I have time I might add an option to customize the wrapping limits.

I remember 80 or 82 characters being a recommendation for the commit message (look at the message of c971ada , github automatically breaks after 82 chars and the header being somewhat close to 54 chars.

Or are you open for changing the default values?

File: index.js (hunk 1 above)
Comment (ngryman) [reply index.js]: Good catch, let's try this then:
```js
const footer = answers.issues.match(/#\d+/g)
  .map(issue => `Closes ${issue}`)
  .join('\n') || []

return [head, body, footer]
  .join('\n\n')
  .trim()
```

File: index.js [hunk 8]
```diff
@@ -1,46 +1,85 @@
...
+const homeDir = require('home-dir')
 const types = require('./lib/types')
 
-function getEmojiChoices(types) {
+const defaultConfig = {
+  types,
+  symbol: false
+}
+
+function getEmojiChoices({types, symbol}) {
```
Comment (ngryman) [inline index.js]: Sorry I didn't even put eslint here, could you just add inner spaces here? 🙏

File: index.js [hunk 9]
```diff
@@ -1,46 +1,85 @@
...
+  return readPkg()
+    .then(({ pkg }) => {
+      const config = (pkg && pkg.config && pkg.config['cz-emoji'])
+      if (config) return config
+
+      return new Promise((resolve, reject) => {
+        fs.readFile(homeDir('.czrc'), 'utf8', (err, content) => {
+          if (err) reject(err)
+          const czrc = JSON.parse(content) || null
+          resolve((czrc && czrc.config && czrc.config['cz-emoji']) || {})
```
Comment (ngryman) [inline index.js]: Thanks for adding the `cz-emoji` section, here you can get rid off `|| {}` as `Object.assign` won't complain about `undefined` values.

Comment (ngryman) [reply index.js]: If you want to, you can also factorize things a bit:
```js
const getConfig = (obj) => obj && obj.config && obj.config['cz-emoji']
```

**Comment by ngryman** [comment]
> Also when I have time I might add an option to customize the wrapping limits.

Sure, that's a good idea. Thanks for proposing this.
A dedicated PR might be better for this, so we can merge this one which already brings lots of stuff 👍

**Comment by nshtg** [comment]
@ngryman Anything left?

**Comment by nshtg** [comment]
Also a small proposal to you: how about adding an .eslintrc so when people contribute the style matches better?

**Comment by ngryman** [comment]
Thanks a lot for your changes, it really makes the project more usable and customisable 👏 🎉.

If you have other ideas they are very welcomed! If you do, could you open an issue before a PR so we quickly discuss about each idea and decide to open the corresponding PR? Also it would be awesome to open one issue per idea so we focus on one thing at a time, it makes things a lot easier 😄

> Also a small proposal to you: how about adding an .eslintrc so when people contribute the style matches better?

Yes I'm gonna add this, I usually do...

**Comment by nshtg** [comment]
@ngryman 
I will do when I have more time.

Btw, would you mind pushing out a release on npm? I want to install this on my remote servers without getting to hacky. ;-)

**Comment by ngryman** [comment]
@nshtg Done! Sorry for the late.

**Pull request merged**